src/esp/gfx/Shader.h

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include <Magnum/GL/AbstractShaderProgram.h>
+
+#include "esp/core/esp.h"
+
+namespace esp {
+namespace gfx {
+
+class Drawable;
+class DrawableGroup;
+class RenderCamera;
+
+// TODO: string type to allow for dynamically added shader types
+enum class ShaderType {
+  /**
+   * Shader program for instance mesh data. See @ref gfx::PrimitiveIDShader,
+   * @ref GenericInstanceMeshData, @ref Mp3dInstanceMeshData, @ref
+   * AssetType::INSTANCE_MESH, @ref loadInstanceMeshData.
+   */
+  INSTANCE_MESH_SHADER = 0,
+
+  /**
+   * Shader program for PTex mesh data. See @ref gfx::PTexMeshShader, @ref
+   * gfx::PTexMeshDrawable, @ref loadPTexMeshData, @ref PTexMeshData.
+   */
+  PTEX_MESH_SHADER = 1,
+
+  /**
+   * Shader program for flat shading with uniform color. Used to render object
+   * identifier or semantic types (e.g. chair, table, etc...). Also the
+   * default shader for assets with unidentified rendering parameters. See
+   * @ref Magnum::Shaders::Flat3D.
+   */
+  COLORED_SHADER = 2,
+
+  /**
+   * Shader program for vertex color shading. Used to render meshes with
+   * per-vertex colors defined.
+   */
+  VERTEX_COLORED_SHADER = 3,
+
+  /**
+   * Shader program for meshes with textured defined.
+   */
+  TEXTURED_SHADER = 4,
+
+  COLORED_SHADER_PHONG = 5,
+  VERTEX_COLORED_SHADER_PHONG = 6,
+  TEXTURED_SHADER_PHONG = 7
+};
+
+struct ShaderConfiguration {
+  ShaderType type = ShaderType::TEXTURED_SHADER;
+  // JSON options;
+};
+
+/**
+ * @brief Shader class
+ */
+class Shader {
+ public:
+  /**
+   * @brief Construct a shader with given configuration
+   */
+  explicit Shader(const ShaderConfiguration& config = {});
+
+  /**
+   * @brief Get the current shader configuration
+   */
+  const ShaderConfiguration& getConfiguration() const { return config_; }
+
+  /**
+   * @brief Update the configuration for this shader
+   * TODO: use return type to signify errors
+   */
+  void setConfiguration(const ShaderConfiguration& config);
+
+  /**
+   * @brief Prepare to draw a @ref DrawableGroup with given @ref RenderCamera
+   *
+   * @return Whether the @ref DrawableGroup can be drawn by the current shader
+   * configuration
+   */
+  bool prepareForDraw(const DrawableGroup& drawables,
+                      const RenderCamera& camera);
```

Comment:
Oops, no idea why I'm actually passing through `DrawableGroup` here. Really this function exists just to set up common shader parameters shared by the whole group, which I don't think can fail. Will update return type to void.

As for the `setConfiguration` idea, any ideas on how to check `isConfigurationCompatible` in a clean way without actually creating a ShaderProgram? I.e. different shader types will have different valid configurations, so checking validity would require creating a new shader via factory (since `Shader` doesn't know about the concrete `ShaderProgram` types), unless I had a separate entry in the lookup table for validation, which I think is overly complicated.

Reply:
I have to admit I was not fully involved in the design discussion for this -- what's the actual definition of a "compatible configuration"? Or how custom can the custom shaders be?

Let's say a custom shader would need an additional buffer texture containing some per-face data to do some fancy shading. Or a PBR shader that additionally needs a metalness texture, or an instanced shader that supplies per-instance matrices via an additional attribute. How would one express that via the configuration? What's the scope of the configuration? Key/value pairs or more? What is a possible use case and what is already out of reach?

I'd say that for a flexible shader management you'd need to:

- have a possibility to call something up-front, before drawing everything that belongs to given shader, for a global setup
- have a possibility to call something before each draw, for a per-mesh setup

In other words, for each new shader, letting users implement their own Drawable that gets used with it, and letting them put things into / operate drawable groups directly. That doesn't need any configuration or configuration validation because users are in charge of that, and ... isn't that kinda already there? I fear this ends up overengineered without really providing the flexibility needed.

Reply:
Thanks for these questions, we definitely need to get the design right before moving further.

The idea is that the "configuration" will specify attributes for shared drawable group setup for any arbitrary type of Shader. A possible use case of this configuration may be to specify type "phong" with attributes numLights, positions, shininess, etc,  or type "flat" with texture flags, vertex colored flags, etc. Now, this `Shader` wrapper class does not know about these underlying shader types, so it must just accept a base type, and the underlying shaderprogram must define what is valid.

So it seems strange that we have the `Shader` wrapper class at all, since we could just allow the client code to configure the underlying `ShaderProgram` directly, before passing it to a DrawableGroup to be used. However, the advantages of this wrapper are:

- The client code doesn't need to know about the underlying shader types being used, then they can just easily pass us a config and we handle shader setup and reuse. Phong implementation could even be swapped out depending on what's registered in the factory.

- The client code doesn't need to store and keep track of their configs (ie. just set and forget, if I want to get it later, and possibly modify it, I can access it with `getConfiguration`

- The configuration can be stored/read in material lighting information in assets (ie. GLtf), instead of always having to be created programmatically.

- The modification/reuse of ShaderPrograms is abstracted. The user shouldn't have to know that adding a light to a phong shader requires me to create a brand new ShaderProgram and set it, but if I just want to move a light I can reuse the existing ShaderProgram. With this wrapper class, they change the configuration in a single uniform way with `setConfiguration`. If necessary, the factories will create a new `ShaderProgram` which supports this new conf. If one already exists (ie. all phong shaders using the same flags and number of lights), we don't need to do anything.

I apologize if any of this isn't clear, its hard to split up the PR's into bite sized chunks that individually make sense, so please let me know if you have questions. Thanks again for the feedback.